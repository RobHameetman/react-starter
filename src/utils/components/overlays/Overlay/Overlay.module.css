.container {
	align-items: center;
	bottom: 0;
	display: flex;
	justify-content: center;
	left: 0;
	position: absolute;
	right: 0;
	top: 0;
	z-index: 1;
}

.container.app {
	height: 100vh;
	pointer-events: none;
	position: fixed;
	width: 100vw;
	z-index: 2147483646;
}

.content {
	align-items: center;
	display: flex;
	height: 100%;
	justify-content: center;
	width: 100%;
	z-index: 2;
}

.app .content {
	z-index: 2147483647;
}

.overlay {
	background-color: var(--overlay-color);
	bottom: 0;
	height: 100%;
	left: 0;
	position: absolute;
	right: 0;
	top: 0;
	width: 100%;
	z-index: -1;
}

.app .overlay {
	pointer-events: auto;
}
