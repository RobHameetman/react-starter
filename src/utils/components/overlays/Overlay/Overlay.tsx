import {
	$FC,
	MouseEvent,
	MouseEventHandler,
	ReactElement,
	useCallback,
	useMemo,
} from 'react';
import { createPortal } from 'react-dom';
import { noop } from '@app/utils/functions/misc/noop';
import { Clickable } from '@app/utils/types/props/Clickable';
import type { Stylable } from '@app/utils/types/props/Stylable';
import styles from './Overlay.module.css';

/**
 * Prop types for `<Overlay />`.
 */
export interface OverlayProps extends Clickable, Stylable {
	/**
	 * [Optional] If `true`, will display the overlay at the app level above all
	 * other content.
	 * @defaultValue - `false`
	 */
	readonly app?: boolean;

	/**
	 * [Optional] Add an animation to the overlay.
	 * @defaultValue - A no-op function.
	 */
	readonly renderOverlay?: ($overlay: ReactElement) => ReactElement;
}

/**
 * An overlay provides contextual information and options in form of an
 * additional layer on top of the current state of the interface. They are
 * either purposefully interruptive like modals or augmenting like popovers and
 * tooltips.
 */
export const Overlay: $FC<OverlayProps> = ({
	app = false,
	children,
	className = '',
	renderOverlay = noop,
	onClick = noop,
}) => {
	const cssOverride = className ? ` ${className}` : '';
	const cssOverlayLevel = app ? ` ${styles.app}` : '';

	const _handleClick = useCallback<MouseEventHandler<HTMLDivElement>>(
		(e) => {
			e.stopPropagation();

			onClick(e);
		},
		[onClick],
	);

	const $overlay = useMemo(
		() => (
			<div
				className={`${styles.overlay}${cssOverride}`}
				role="overlay"
				onClick={_handleClick}
			/>
		),
		[_handleClick, cssOverride],
	);

	const $jsx = useMemo(
		() => (
			<div className={`${styles.container}${cssOverlayLevel}`}>
				<div className={styles.content}>{children}</div>
				{renderOverlay ? renderOverlay($overlay) : $overlay}
			</div>
		),
		[$overlay, children, cssOverlayLevel, renderOverlay],
	);

	return app && appOverlayRef.current
		? createPortal($jsx, appOverlayRef.current)
		: $jsx;
};
