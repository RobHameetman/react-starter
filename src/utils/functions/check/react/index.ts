export * from './isChangeEvent';
export * from './isDragEvent';
export * from './isEventHandler';
export * from './isFocusEvent';
export * from './isFocusEventHandler';
export * from './isKeyboardEvent';
export * from './isMouseEvent';
export * from './isMutableRefObject';
export * from './isPointerEvent';
export * from './isReactElement';
export * from './isScrollEvent';
export * from './isSyntheticEvent';
export * from './isUIEvent';
export * from './isUIEventHandler';
