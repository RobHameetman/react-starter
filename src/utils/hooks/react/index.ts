export * from './useControlProp';
export * from './useEnterEvents';
export * from './useEscapeEvents';
export * from './useFocusEvents';
export * from './usePressEvents';
export * from './usePropsWithChildren';
export * from './useSemanticAsProp';
export * from './useSpaceEvents';
export * from './useTabEvents';
