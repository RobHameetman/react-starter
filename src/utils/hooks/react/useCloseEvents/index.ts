export * from './useCloseEvents';
export { default } from './useCloseEvents';
