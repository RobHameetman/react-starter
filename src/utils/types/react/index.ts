export * from './CC';
export * from './CompoundComponent';
export * from './HandlerPropName';
export * from './HandlerPropsOf';
export * from './PropsWithoutHandlers';
