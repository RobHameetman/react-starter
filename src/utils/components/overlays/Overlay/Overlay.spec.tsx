import { render } from '@testing-library/react';
import { Overlay } from './Overlay';

describe('Overlay', () => {
	it('should render correctly', () => {
		expect(() => render(<Overlay />)).not.toThrowError();
	});

	it('should not regress', () => {
		expect(() => render(<Overlay />)).toMatchSnapshot();
	});
});
