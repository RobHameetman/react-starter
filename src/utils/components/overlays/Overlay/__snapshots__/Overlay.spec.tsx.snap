// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should not regress 1`] = `"<div class=\\"container\\"><div class=\\"content\\"></div><div class=\\"overlay\\" typeof=\\"overlay\\"></div></div>"`;
